### PR TITLE
frontend: Allow opt out of EnsureRevision when making a comparison query

### PR DESCRIPTION
**Pull request by ryanslade**
Add an optional `FetchMissing` field to the `comparison` query. The
default is true which respects the previous default behaviour.

If set to false then we will not attempt to fetch revisions that are
missing.

Part of: https://github.com/sourcegraph/sourcegraph/issues/11654

### CODE REVIEW

File: cmd/frontend/backend/repos_vcs.go [hunk 1]
```diff
@@ -197,7 +197,7 @@ func (s *repos) GetCommit(ctx context.Context, repo *types.Repo, commitID api.Co
 		}
 		return grepo.URL, nil
 	}
-	return git.GetCommit(ctx, *gitserverRepo, remoteURLFunc, commitID)
+	return git.GetCommit(ctx, *gitserverRepo, remoteURLFunc, commitID, git.ResolveRevisionOptions{})
```
Comment (ryanslade) [inline cmd/frontend/backend/repos_vcs.go:200]: I chose a value here instead of a pointer to avoid nil checks and use the zero value.

**Comment by tsenart** [comment]
Tests?

**Comment by ryanslade** [comment]
@tsenart PTAL

File: cmd/frontend/backend/repos_vcs.go (hunk 1 above)
Comment (tsenart) [reply cmd/frontend/backend/repos_vcs.go:200]: It introduces an inconsistency with the other methods like `git.ResolveRevision` though, so I'd rather pass a pointer in.

File: internal/vcs/git/commits.go [hunk 2]
```diff
@@ -65,7 +68,7 @@ func getCommit(ctx context.Context, repo gitserver.Repo, remoteURLFunc func() (s
 		return nil, err
 	}
 
-	commits, err := commitLog(ctx, repo, CommitsOptions{Range: string(id), N: 1, RemoteURLFunc: remoteURLFunc})
+	commits, err := commitLog(ctx, repo, CommitsOptions{Range: string(id), N: 1, RemoteURLFunc: remoteURLFunc, NoEnsureRevision: opt.NoEnsureRevision})
```
Comment (tsenart) [inline internal/vcs/git/commits.go:71]: Long line is loooong.

File: internal/vcs/git/commits_test.go [hunk 3]
```diff
@@ -27,19 +25,39 @@ func TestRepository_GetCommit(t *testing.T) {
 		Parents:   []api.CommitID{"ea167fe3d76b1e5fd3ed8ca44cbd2fe3897684f8"},
 	}
 	tests := map[string]struct {
-		repo       gitserver.Repo
-		id         api.CommitID
-		wantCommit *Commit
+		repo             gitserver.Repo
+		id               api.CommitID
+		wantCommit       *Commit
+		NoEnsureRevision bool
```
Comment (tsenart) [inline internal/vcs/git/commits_test.go:31]:
```suggestion
		noEnsureRevision bool
```

File: internal/vcs/git/commits.go [hunk 4]
```diff
@@ -49,14 +49,17 @@ type CommitsOptions struct {
 	// gitserver doesn't already contain a clone of the repository or if the
 	// commit must be fetched from the remote.
 	RemoteURLFunc func() (string, error)
+
+	// When true we opt out of attempting to fetch missing revisions
+	NoEnsureRevision bool
```
Comment (unknwon) [inline internal/vcs/git/commits.go:54]: nit:
```suggestion
	// When true we opt out of attempting to fetch missing revisions.
	SkipEnsureRevision bool
```

Comment (ryanslade) [reply internal/vcs/git/commits.go:54]: We use this naming in other parts of the code so I wanted to stay consistent

**Pull request merged**